[Inference API] Add Jina AI API to do inference for Embedding and Rerank models

JoanFM opened the pull request:
This is the code to add Jina AI as a plugin to do inference for embeddigns and reranking

JoanFM committed:
feat: add basic code under inference/external

JoanFM committed:
feat: add basic code under inference/services

JoanFM committed:
code is compiling

JoanFM committed:
add tests and apply pre commit style

JoanFM committed:
compiles and some tests pass

JoanFM committed:
pre commit style

JoanFM committed:
most of tests pass

JoanFM committed:
most of tests pass and precommit

JoanFM committed:
tests are passing

JoanFM committed:
more tests passing

JoanFM committed:
finished JinaAIServiceTests

JoanFM committed:
Merge branch 'main' into feat-add-jina-ai

JoanFM committed:
fix merge conflicts

JoanFM committed:
name of service lowercase

JoanFM committed:
fix inputs key

JoanFM committed:
Merge branch 'main' into feat-add-jina-ai

elasticsearchmachine commented:
Pinging @elastic/ml-core (Team:ML)

timgrein commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/jinaai/JinaAIResponseHandler.java:
```diff
@@ -0,0 +1,98 @@
+        checkForEmptyBody(throttlerManager, logger, request, result);
+    }
+
+    @Override
+    public boolean canHandleStreamingResponses() {
+        return false;
+    }
+
+    @Override
+    public InferenceServiceResults parseResult(Request request, Flow.Publisher<HttpResult> flow) {
+        // TODO (JoanFM): Check this logic
+        return new StreamingChatCompletionResults(null);
+    }
```
```suggestion
```

I think you can remove these two overrides.

The top interface [ResponseHandler](https://github.com/elastic/elasticsearch/blob/d09d57db802ecba45195abfd1b38c12818bbe537/x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/http/retry/ResponseHandler.java#L45-L74) in the inheritance chain provides default implementations for both `canHandleStreamingResponses` and `parseResult` handling providers, which do not handle streaming responses.

timgrein commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/jinaai/JinaAIResponseHandler.java:
```diff
@@ -0,0 +1,98 @@
+ *
+ */
+public class JinaAIResponseHandler extends BaseResponseHandler {
+    static final String INPUTS_ARRAY_TOO_LARGE_MESSAGE_MATCHER = "is larger than the largest allowed size";
```
```suggestion
    static final String INPUTS_ARRAY_TOO_LARGE_MESSAGE_MATCHER = "input array is larger than the largest allowed size";
```

nit: I think we can add _what_ is larger than the largest allowed size

timgrein commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/jinaai/JinaAIResponseHandler.java:
```diff
@@ -0,0 +1,98 @@
+        }
+    }
+
+    private static boolean isTextsArrayTooLarge(HttpResult result, int statusCode) {
+
+        if (statusCode == 400) {
+            var errorEntity = JinaAIErrorResponseEntity.fromResponse(result);
+
+            return errorEntity != null && errorEntity.getErrorMessage().contains(INPUTS_ARRAY_TOO_LARGE_MESSAGE_MATCHER);
+        }
+
+        return false;
+    }
```
From a JinaAI perspective it probably makes sense to handle these specific errors, but from Elastic's perspective it probably makes more sense to stay generic and do not check concrete error messages as these could change at any time without us noticing. I guess your API returns anyway a useful error? (assuming it as you're checking for one :) )


I suggest to remove the error message specific response handling. WDYT @jonathan-buttner ?

JoanFM replied:
I saw it from `Cohere`. But I agree with the thing that we should not depend on this

timgrein commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/request/jinaai/JinaAIEmbeddingsRequestEntity.java:
```diff
@@ -0,0 +1,68 @@
+    }
+
+    // default for testing
+    static String covertToString(InputType inputType) {
```
```suggestion
    static String convertToString(InputType inputType) {
```

timgrein commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/response/jinaai/JinaAIRankedResponseEntity.java:
```diff
@@ -0,0 +1,159 @@
+import static org.elasticsearch.xpack.inference.external.response.XContentUtils.moveToFirstToken;
+import static org.elasticsearch.xpack.inference.external.response.XContentUtils.positionParserAtTokenAfterField;
+
+public class JinaAIRankedResponseEntity {
```
```suggestion
public class JinaAIRerankResponseEntity {
```

I think this should be called `[...]Rerank[...]` instead of `[...]Ranked[...]` for consistency reasons.

timgrein commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/response/jinaai/JinaAIRankedResponseEntity.java:
```diff
@@ -0,0 +1,159 @@
+    private static final Logger logger = LogManager.getLogger(JinaAIRankedResponseEntity.class);
+
+    /**
+     * //TODO(JoanFM): Edit documentation
```
Marker: is this addressed?

timgrein commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/response/jinaai/JinaAIRankedResponseEntity.java:
```diff
@@ -0,0 +1,159 @@
+            XContentParser.Token token = jsonParser.currentToken();
+            ensureExpectedToken(XContentParser.Token.START_OBJECT, token, jsonParser);
+
+            positionParserAtTokenAfterField(jsonParser, "results", FAILED_TO_FIND_FIELD_TEMPLATE); // TODO error message
```
Marker: is this addressed?

timgrein commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/jinaai/JinaAIService.java:
```diff
@@ -0,0 +1,370 @@
+
+    /**
+     * Return the default similarity measure for the embedding type.
+     * JinaAI embeddings are normalized to unit vectors therefor Dot
```
```suggestion
     * JinaAI embeddings are normalized to unit vectors therefore Dot
```

timgrein commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/jinaai/JinaAIService.java:
```diff
@@ -0,0 +1,370 @@
+
+    @Override
+    public TransportVersion getMinimalSupportedVersion() {
+        return TransportVersions.V_8_15_0;
```
I think you need a new `TransportVersion` here, which should be called something like `JINA_AI_INTEGRATION_ADDED`. You need to add it at the bottom of [TransportVersions](https://github.com/elastic/elasticsearch/blob/d09d57db802ecba45195abfd1b38c12818bbe537/server/src/main/java/org/elasticsearch/TransportVersions.java#L140).

timgrein commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/jinaai/JinaAIService.java:
```diff
@@ -0,0 +1,370 @@
+        return TransportVersions.V_8_15_0;
+    }
+
+    @Override
+    public Set<TaskType> supportedStreamingTasks() {
+        return Set.of();
+    }
```
I think this override is not necessary. `InferenceService` already returns `Set.of()`

timgrein commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/jinaai/JinaAIServiceSettings.java:
```diff
@@ -0,0 +1,180 @@
+public class JinaAIServiceSettings extends FilteredXContentObject implements ServiceSettings, JinaAIRateLimitServiceSettings {
+
+    public static final String NAME = "jinaai_service_settings";
+    public static final String OLD_MODEL_ID_FIELD = "model";
```
I don't think we need the `OLD_MODEL_ID_FIELD`. This is a relict from the past when we used `model` instead of `model_id`. Every integration after the introduction of the new model id shouldn't handle the old model id. We just handle it for integrations, where we need backwards-compatibility.

timgrein commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/jinaai/JinaAIServiceSettings.java:
```diff
@@ -0,0 +1,180 @@
+        uri = createOptionalUri(in.readOptionalString());
+        modelId = in.readOptionalString();
+
+        if (in.getTransportVersion().onOrAfter(TransportVersions.V_8_15_0)) {
```
This transport version check can be removed as soon as you introduced the new transport version mentioned above. This is just needed for providers, which were introduced before rate limiting.

timgrein commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/jinaai/JinaAIServiceSettings.java:
```diff
@@ -0,0 +1,180 @@
+        out.writeOptionalString(uriToWrite);
+        out.writeOptionalString(modelId);
+
+        if (out.getTransportVersion().onOrAfter(TransportVersions.V_8_15_0)) {
```
Same as above

timgrein commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/jinaai/embeddings/JinaAIEmbeddingsServiceSettings.java:
```diff
@@ -0,0 +1,162 @@
+
+    @Override
+    public TransportVersion getMinimalSupportedVersion() {
+        return TransportVersions.V_8_13_0;
```
This should also use the new transport version

timgrein commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/jinaai/embeddings/JinaAIEmbeddingsTaskSettings.java:
```diff
@@ -0,0 +1,183 @@
+
+    @Override
+    public TransportVersion getMinimalSupportedVersion() {
+        return TransportVersions.V_8_13_0;
```
Same as for the other comments about transport version

timgrein commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/jinaai/rerank/JinaAIRerankServiceSettings.java:
```diff
@@ -0,0 +1,122 @@
+
+    @Override
+    public TransportVersion getMinimalSupportedVersion() {
+        return TransportVersions.V_8_14_0;
```
Same as for other comments about transport version

timgrein commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/jinaai/rerank/JinaAIRerankTaskSettings.java:
```diff
@@ -0,0 +1,166 @@
+
+    @Override
+    public TransportVersion getMinimalSupportedVersion() {
+        return TransportVersions.V_8_14_0;
```
Same as for the other comments about transport version

timgrein reviewed (commented):
Thank you for your contribution! 👏 

Excited to see JinaAI becoming another provider in Elastic's inference APIs. I gave your PR a first review pass (excluding tests for now) - I'll give it another review after these comments are addressed. Thanks!

davidkyle committed:
Add changelog

davidkyle commented:
@elasticmachine test this please

davidkyle commented:
For some reason the bot doesn't automatically create a changelog for external contributors. I pushed an entry to in https://github.com/elastic/elasticsearch/pull/118652/commits/7915374d8763de5dd8f463750d48dceee0b801b5

JoanFM committed:
apply code comments

JoanFM committed:
Merge branch 'feat-add-jina-ai' of https://github.com/JoanFM/elasticsearch into feat-add-jina-ai

JoanFM commented:
Hey @timgrein ,

Thanks a lot for the review.

I addressed all the comments. However, there is this test not passing related to something with this `transport` version.

Could you take a look? I could not find the test itself, I think it comes from some template class.

org.elasticsearch.xpack.inference.services.jinaai.rerank.JinaAIRerankServiceSettingsTests.testBwcSerialization


Thanks a lot

timgrein commented on x-pack/plugin/inference/src/test/java/org/elasticsearch/xpack/inference/services/jinaai/rerank/JinaAIRerankServiceSettingsTests.java:
```diff
@@ -0,0 +1,99 @@
+
+    @Override
+    protected JinaAIRerankServiceSettings mutateInstanceForVersion(JinaAIRerankServiceSettings instance, TransportVersion version) {
+        if (version.before(TransportVersions.JINA_AI_INTEGRATION_ADDED)) {
+            // We always default to the same rate limit settings, if a node is on a version before rate limits were introduced
+            return new JinaAIRerankServiceSettings(
+                new JinaAIServiceSettings(
+                    instance.getCommonSettings().uri(),
+                    instance.modelId(),
+                    JinaAIServiceSettings.DEFAULT_RATE_LIMIT_SETTINGS
+                )
+            );
+        }
+        return instance;
```
```suggestion
        return instance;
```

I think you can simply return `instance` here without any transport version specific mutations. As elasticsearch nodes with a version before `TransportVersions.JINA_AI_INTEGRATION_ADDED` won't have the JinaAI integration we're 100% certain that the JinaAI integration supports rate limiting.

timgrein reviewed (commented):
Thanks for addressing the comments! Left an additional comment with regards to the failing test case. Let me know, if that works for you

JoanFM committed:
apply suggestions from code review

Co-authored-by: Tim Grein <tim@4greins.de>

JoanFM commented:
Seems to solve the issue indeed

JoanFM committed:
Merge branch 'main' into feat-add-jina-ai

timgrein commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/action/jinaai/JinaAIActionCreator.java:
```diff
@@ -0,0 +1,60 @@
+    private final ServiceComponents serviceComponents;
+
+    public JinaAIActionCreator(Sender sender, ServiceComponents serviceComponents) {
+        // TODO Batching - accept a class that can handle batching
```
Just for my understanding of this `TODO`: what do you mean with batching here? Batching a set of requests over some window (time, amount of requests etc.) into one request and sending them to your API once?

JoanFM replied:
to be honest, this remains here from the other providers where I took inspiration from.

timgrein commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/action/jinaai/JinaAIActionCreator.java:
```diff
@@ -0,0 +1,60 @@
+            overriddenModel.getServiceSettings().getCommonSettings().uri(),
+            "JinaAI embeddings"
+        );
+        // TODO - Batching pass the batching class on to the JinaAIEmbeddingsRequestManager
```
Same question as above

timgrein reviewed (commented):
Glad to hear tests work now! 🙏 

Added one question around the batching TODOs.

JoanFM committed:
Merge branch 'main' into feat-add-jina-ai

JoanFM committed:
Merge branch 'main' into feat-add-jina-ai

timgrein commented on x-pack/plugin/inference/src/test/java/org/elasticsearch/xpack/inference/services/jinaai/embeddings/JinaAIEmbeddingsServiceSettingsTests.java:
```diff
@@ -0,0 +1,227 @@
+    public static JinaAIEmbeddingsServiceSettings createRandom() {
+        SimilarityMeasure similarityMeasure = null;
+        Integer dims = null;
+        var isTextEmbeddingModel = randomBoolean();
+        if (isTextEmbeddingModel) {
+            similarityMeasure = SimilarityMeasure.DOT_PRODUCT;
+            dims = 1024;
+        }
```
```suggestion
        similarityMeasure = SimilarityMeasure.DOT_PRODUCT;
        dims = 1024;
```

`JinaAIEmbeddingsServiceSettings` always refers to the `text_embedding` task type, right? Then we don't need to test the case, where it's not a text embedding model (as it always is one AFAIU).

timgrein commented on x-pack/plugin/inference/src/test/java/org/elasticsearch/xpack/inference/services/jinaai/embeddings/JinaAIEmbeddingsTaskSettingsTests.java:
```diff
@@ -0,0 +1,189 @@
+        var initialSettings = createRandom();
+        var newSettings = createRandom();
+        Map<String, Object> newSettingsMap = new HashMap<>();
+        if (newSettings.getInputType() != null) {
+            newSettingsMap.put(JinaAIEmbeddingsTaskSettings.INPUT_TYPE, newSettings.getInputType().toString());
+        }
+        JinaAIEmbeddingsTaskSettings updatedSettings = (JinaAIEmbeddingsTaskSettings) initialSettings.updatedTaskSettings(
+            Collections.unmodifiableMap(newSettingsMap)
+        );
+        if (newSettings.getInputType() == null) {
+            assertEquals(initialSettings.getInputType(), updatedSettings.getInputType());
+        } else {
+            assertEquals(newSettings.getInputType(), updatedSettings.getInputType());
```
I would suggest to separate these two test cases into:

- `testUpdatedTaskSettings_NotUpdated_UseInitialSettings`
- `testUpdatedTaskSettings_Updated_UseNewSettings`

This makes it a bit easier to reason about potential failures IMO + the test is much easier to read.

timgrein reviewed (commented):
Thanks for the very extensive testing! 🥇 

Added some minor comments. I'll do some e2e-tests with your branch and then I'm happy to approve! Thanks again for your responsiveness and the iterations, great contribution.

JoanFM committed:
apply suggestions from code review

Co-authored-by: Tim Grein <tim@4greins.de>

JoanFM committed:
apply more suggestions

JoanFM commented:
Hey @timgrein ,

I applied all suggestions

davidkyle commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/response/jinaai/JinaAIEmbeddingsResponseEntity.java:
```diff
@@ -0,0 +1,111 @@
+     *          "object": "embedding",
+     *          "embedding": [
+     *              -0.009327292,
+     *              .... (1536 floats total for ada-002)
```
```suggestion
```

JoanFM committed:
update x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/response/jinaai/JinaAIEmbeddingsResponseEntity.java

Co-authored-by: David Kyle <david.kyle@elastic.co>

timgrein commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/jinaai/rerank/JinaAIRerankServiceSettings.java:
```diff
@@ -0,0 +1,122 @@
+    public static JinaAIRerankServiceSettings fromMap(Map<String, Object> map, ConfigurationParseContext context) {
+        ValidationException validationException = new ValidationException();
+        var commonServiceSettings = JinaAIServiceSettings.fromMap(map, context);
+        // We need to extract/remove those fields to avoid unknown service settings errors
+        extractSimilarity(map, ModelConfigurations.SERVICE_SETTINGS, validationException);
+        removeAsType(map, DIMENSIONS, Integer.class);
+        removeAsType(map, MAX_INPUT_TOKENS, Integer.class);
```
I think this is not needed. I guess you've looked at `CohereRerankServiceSettings`? These rerank service settings contained text embedding service settings in the past (which was wrong), therefore we had to do this dance in `CohereRerankServiceSettings` to maintain backwards-compability while fixing the bug.

JoanFM replied:
Yes I noticed that and did proper refactoring of my Service Settings inheritance trees. However I just mantained it in case it was necessary. I can remove these

timgrein replied:
Thanks 🙏

timgrein reviewed (commented):
Just one additional small thing I've noticed, which is not needed and could be confusing in the future.

JoanFM committed:
remove unneeded code in rerank service settings

davidkyle commented on x-pack/plugin/inference/src/test/java/org/elasticsearch/xpack/inference/services/jinaai/rerank/JinaAIRerankTaskSettingsTests.java:
```diff
@@ -0,0 +1,132 @@
+        assertThat(thrownException.getMessage(), containsString("field [top_n] is not of the expected type"));
+    }
+
+    public void UpdatedTaskSettings_WithEmptyMap_ReturnsSameSettings() {
```
```suggestion
    public void testUpdatedTaskSettings_WithEmptyMap_ReturnsSameSettings() {
```

Tests need to start with `test`

davidkyle commented on x-pack/plugin/inference/src/test/java/org/elasticsearch/xpack/inference/services/jinaai/rerank/JinaAIRerankModelTests.java:
```diff
@@ -0,0 +1,74 @@
+import org.elasticsearch.xpack.inference.services.jinaai.JinaAIServiceSettings;
+import org.elasticsearch.xpack.inference.services.settings.DefaultSecretSettings;
+
+public class JinaAIRerankModelTests extends ESTestCase {
```
I don't see any tests in this file. Are the various `createModel(...)` methods used

JoanFM replied:
they are used in different tests. I took the idea from Cohere

jonathan-buttner replied:
nit: If there are no tests we can probably remove the `extends ESTestCase`

davidkyle reviewed (commented):
Looks great thanks for the PR @JoanFM. I commented on some very minor points

davidkyle commented:
@elasticmachine test this please

JoanFM committed:
Merge branch 'main' into feat-add-jina-ai

JoanFM committed:
apply suggestions from code review

Co-authored-by: David Kyle <david.kyle@elastic.co>

jonathan-buttner commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/jinaai/embeddings/JinaAIEmbeddingsServiceSettings.java:
```diff
@@ -0,0 +1,162 @@
+public class JinaAIEmbeddingsServiceSettings extends FilteredXContentObject implements ServiceSettings {
+    public static final String NAME = "jinaai_embeddings_service_settings";
+
+    public static JinaAIEmbeddingsServiceSettings fromMap(Map<String, Object> map, ConfigurationParseContext context) {
```
Did we want to support the `embedding_type` field? I believe Elasticsearch supports floats and bits, I'm not sure if that maps to ubinary or binary, I'd have to look.

JoanFM replied:
I believe the embedding types returned by Jina API will not be of much use here.

jonathan-buttner commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/jinaai/rerank/JinaAIRerankTaskSettings.java:
```diff
@@ -0,0 +1,166 @@
+     */
+    public static JinaAIRerankTaskSettings of(JinaAIRerankTaskSettings originalSettings, JinaAIRerankTaskSettings requestTaskSettings) {
+        return new JinaAIRerankTaskSettings(
+            requestTaskSettings.getTopNDocumentsOnly() != null
```
nit: I think we could simplify these with `Objects.requireNonNullElse`

JoanFM replied:
Could u elaborate how?

jonathan-buttner replied:
Something like this:

```
Objects.requireNonNullElse(requestTaskSettings.getTopNDocumentsOnly(), originalSettings.getTopNDocumentsOnly());
```

You could do that for each constructor parameter instead of the ternary operators.

JoanFM replied:
But this requires the default not to be nullable, and this is not the case here. I am keeping as it is now.

jonathan-buttner replied:
Ah you're right! Sorry about that, I forgot that it validates that the second parameter is non-null 👍

jonathan-buttner reviewed (changes requested):
Great work! I left a few comments some nits and some changes.

A few of the comments might already be address (around the TODOs) feel free to resolve those comments if so 👍

JoanFM commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/http/sender/JinaAIRequestManager.java:
```diff
@@ -0,0 +1,28 @@
+        public static RateLimitGrouping of(JinaAIModel model) {
+            Objects.requireNonNull(model);
+
+            return new RateLimitGrouping(model.apiKey().hashCode());
```
yes

JoanFM commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/jinaai/JinaAIServiceSettings.java:
```diff
@@ -0,0 +1,162 @@
+        ValidationException validationException = new ValidationException();
+
+        String url = extractOptionalString(map, URL, ModelConfigurations.SERVICE_SETTINGS, validationException);
+        URI uri = convertToUri(url, URL, ModelConfigurations.SERVICE_SETTINGS, validationException);
```
what do u propose?

jonathan-buttner replied:
If you'd like to make the changes, then doing something like I mentioned in those likes is probably preferred. If you don't have the time, no worries.

JoanFM replied:
I do not get exactly the difference, I would appreciate if it can be done later

JoanFM commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/jinaai/embeddings/JinaAIEmbeddingsTaskSettings.java:
```diff
@@ -0,0 +1,183 @@
+        InputType.INGEST,
+        InputType.SEARCH,
+        InputType.CLASSIFICATION,
+        InputType.CLUSTERING
```
`separation` is our translation to `clustering`

jonathan-buttner replied:
Oh right, sorry. What about `text-matching`?

JoanFM replied:
Since it is not handled by ES, I do not think it makes much sense here

jonathan-buttner replied:
Sounds good!

JoanFM commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/jinaai/embeddings/JinaAIEmbeddingsServiceSettings.java:
```diff
@@ -0,0 +1,162 @@
+        var commonServiceSettings = JinaAIServiceSettings.fromMap(map, context);
+        SimilarityMeasure similarity = extractSimilarity(map, ModelConfigurations.SERVICE_SETTINGS, validationException);
+        Integer dims = removeAsType(map, DIMENSIONS, Integer.class);
+        Integer maxInputTokens = removeAsType(map, MAX_INPUT_TOKENS, Integer.class);
```
I am not sure about these

JoanFM commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/jinaai/JinaAIService.java:
```diff
@@ -0,0 +1,364 @@
+        if (model instanceof JinaAIEmbeddingsModel embeddingsModel) {
+            var serviceSettings = embeddingsModel.getServiceSettings();
+            var similarityFromModel = serviceSettings.similarity();
+            var similarityToUse = similarityFromModel == null ? defaultSimilarity() : similarityFromModel;
```
how would these changes look like?

jonathan-buttner replied:
Take a look here: https://github.com/elastic/elasticsearch/blob/32ddbb3449d19a0970b96eefe960d4ab006357fc/x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/openai/embeddings/OpenAiEmbeddingsServiceSettings.java#L102

Basically you'd do a similar call to parsing the similarity field optionally, if it wasn't present then you'd default it there. That way it'd aways be set.

davidkyle commented:
@elasticmachine test this please

davidkyle commented on docs/changelog/118652.yaml:
```diff
@@ -0,0 +1,5 @@
+pr: 118652
+summary: [Inference API] Add Jina AI API to do inference for Embedding and Rerank models
```
```suggestion
summary: Add Jina AI API to do inference for Embedding and Rerank models
```

davidkyle committed:
Update docs/changelog/118652.yaml

JoanFM committed:
apply comments

JoanFM committed:
Merge branch 'feat-add-jina-ai' of https://github.com/JoanFM/elasticsearch into feat-add-jina-ai

JoanFM commented:
I have applied the suggestions, special focus on making `model` non-Nullable in every layer

JoanFM committed:
Merge branch 'main' into feat-add-jina-ai

davidkyle committed:
fix check style

davidkyle commented:
@elasticmachine test this please

davidkyle commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/jinaai/JinaAIServiceSettings.java:
```diff
@@ -0,0 +1,164 @@
+    public static final String MODEL_ID = "model_id";
+    private static final String DEFAULT_MODEL_ID = "jina-embeddings-v3";
+    private static final Logger logger = LogManager.getLogger(JinaAIServiceSettings.class);
+    public static final RateLimitSettings DEFAULT_RATE_LIMIT_SETTINGS = new RateLimitSettings(2_000); // https://jina.ai/contact-sales/#rate-limit
```
```suggestion
    // See https://jina.ai/contact-sales/#rate-limit
    public static final RateLimitSettings DEFAULT_RATE_LIMIT_SETTINGS = new RateLimitSettings(2_000); 
```

check style is complaining about the long line

davidkyle committed:
Update x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/jinaai/JinaAIServiceSettings.java

davidkyle commented:
@elasticmachine test this please

Commit:
[CI] Auto commit changes from spotless

JoanFM committed:
Merge branch 'main' into feat-add-jina-ai

davidkyle commented:
@elasticmachine test this please

jonathan-buttner commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/jinaai/JinaAIServiceSettings.java:
```diff
@@ -83,7 +86,7 @@ public JinaAIServiceSettings(StreamInput in) throws IOException {
 
     // should only be used for testing, public because it's accessed outside of the package
     public JinaAIServiceSettings() {
```
It doesn't look like this constructor is used, I think I'd prefer if we removed it that way we don't accidentally try to create a default service settings with a default model id.

jonathan-buttner reviewed (commented):
Looks great! One last request about removing an unused constructor.

JoanFM committed:
remove empty constructor

jonathan-buttner reviewed (approved):
Thanks for the changes! 🚢

davidkyle commented:
@elasticmachine test this please

timgrein reviewed (approved):
Nice, thanks for addressing all comments! 🚢

JoanFM committed:
Merge branch 'main' into feat-add-jina-ai

davidkyle committed:
fix test

davidkyle commented:
@elasticmachine test this please

davidkyle committed:
Merge branch 'main' into feat-add-jina-ai

davidkyle commented:
@elasticmachine test this please

JoanFM committed:
Merge branch 'main' into feat-add-jina-ai

Commit:
[CI] Auto commit changes from spotless

maxhniebergall committed:
Merge branch 'main' into feat-add-jina-ai

maxhniebergall commented:
@elasticmachine test this please

JoanFM committed:
Merge branch 'main' into feat-add-jina-ai

JoanFM committed:
Merge branch 'main' into feat-add-jina-ai

davidkyle committed:
Merge branch 'main' into feat-add-jina-ai

davidkyle commented:
@elasticmachine test this please

Commit:
[CI] Auto commit changes from spotless

davidkyle committed:
Merge branch 'main' into feat-add-jina-ai

elasticsearchmachine commented:
## 💔 Backport failed
| Status | Branch | Result |
|:------:|:------:|:------:|
| ❌ |  8.x  | Commit could not be cherrypicked due to conflicts |

You can use [sqren/backport](https://github.com/sqren/backport) to manually backport by running `backport --upstream elastic/elasticsearch --pr 118652`
